### PR TITLE
Add Drawer Personalization

### PR DESCRIPTION
**Pull request by werererer**
#### What is it?
- [ ] Bug fix (user facing)
- [x] Feature (user facing)
- [x] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- seperate Drawer stuff from MainActivityClass 
- customize Drawer/Panel

#### Fixes the following issue(s)

- https://github.com/TeamNewPipe/NewPipe/issues/4081

#### Test Apk
[newpipeDrawerupdateApk.zip](https://github.com/TeamNewPipe/NewPipe/files/5047706/newpipeDrawerupdateApk.zip)

#### Agreement
- [X] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

This is my first PR that had android studio involved. It incoporates a feature to Configure the left Panel/Drawer. This can be done under settings/Content/Content of main page.

I also seperated Drawerlogic from the main Activity class. The Drawerlogic is now in the DrawerFragment class.

### CODE REVIEW

**Comment by opusforlife2** [comment]
Maybe it should be under Appearance? Worth discussion, I think.

**Comment by werererer** [comment]
I think so since changing the drawer changes the Appearance

**Comment by TobiGr** [comment]
Thank you for your contribution  @werererer!
Please update Gradle in a separate PR.
PS: you can tick checkboxes by using `- [x]` (no spaces between the brackets and the x)

**Comment by opusforlife2** [comment]
> PS: you can tick checkboxes by using - [x] (no spaces between the brackets and the x)

Or, far easier, you can just tick the boxes by clicking on them after you've already created the post.

**Comment by werererer** [comment]
What do you mean with Gradle in a separate PR @TobiGr.
Sorry I am not really a Java main so I don't know what you meant with "Please update Gradle in a separate PR".
PS: thanks for the tipps :D

File: build.gradle [hunk 1]
```diff
@@ -7,7 +7,7 @@ buildscript {
         google()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.6.3'
+        classpath 'com.android.tools.build:gradle:4.0.1'
```
Comment (wb9688) [inline build.gradle:10]: Here

File: gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,6 +1,6 @@
-#Fri May 01 19:39:41 CEST 2020
+#Sun Aug 09 14:37:38 CEST 2020
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.1.1-all.zip
```
Comment (wb9688) [inline gradle/wrapper/gradle-wrapper.properties:6]: And here

**Comment by opusforlife2** [comment]
Since you used Android Studio, you must have gotten that update prompt "Update Gradle Plugin". That is some stupid setting that Google introduced which automatically screws with your current projects, namely Newpipe. When you get that prompt, just close it. Or, now that you have updated already, just undo that change from the build.gradle file as wb9688 said.

**Comment by werererer** [comment]
Ok

**Comment by werererer** [comment]
Ok fixed it

**Comment by werererer** [comment]
Some tips for reviewing (as i didn't split up the pull request sry. After some research I know now that it should be done :/):
- a lot of functions moved from Mainactivity to Drawefragment
- thats why i delete that much there
- AddSectionDialog, ChooseSectionsFragment, Section, SectionsJsonHelper and SectionsManager have the exact same task as the Tab counterpart. 
- except that they are for the drawer
- I copied a lot of the code from other classes infact and altert it then according to the need
- we may even let the sections and tabs classes extend the same interface/class

**Comment by MD77MD** [comment]
>we may even let the sections and tabs classes extend the same interface/class (as they are so similar)
@werererer could you take a look at this: 
#3825 Reorder and manage newPipe tabs from home screen 

while your working on this, I wish if customizing can be done directly from drawer, without the need to dig into settings. 
from my point of view, newPipe is always expanding and adding new sources for content (YouTube, SoundCloud...). at the same time, adding more personalization options like this PR as well as functionalities like pinning channels & playlists to taps. with time customizing drawer and tabs for each service can become a hassle. each time a user needs to change something, the users to dig deep and settings.

In my opinion, the easiest way to solve this is to use pop-up menu (similarly to video pop-up menu). the menu should have all the function needed (remove, add, reorder...etc). similarly, the menu can be activated by long pressing on the item or tab you want to change. 

If customization could be done this way, it would make personalization of all the services so much easier. I know this needs more effort from you but it's worth it. 

newPipe deserves perfection 💖

**Comment by werererer** [comment]
good idea i will work on it later :)

**Comment by werererer** [comment]
I looked at it and i realized that this would require way more work than expected. I also don't really have the knowledge to implement this :/(it would require a new drawer and panel that has a recyclerview in it?). As for this PR it may be overkill :/

**Comment by wb9688** [comment]
I think just having it inside the settings is OK

**Comment by werererer** [comment]
ok I think everything works now I don't plan to make any greater changes in this PR (if I didn't introduce bugs)

**Comment by wb9688** [comment]
Could you rebase on dev and fix potential Checkstyle issues? I'll take a look at the UI and code later.

**Comment by werererer** [comment]
how do i rebase? (sry still relatively new to github)

**Comment by werererer** [comment]
nevermind working on it now

**Comment by wb9688** [comment]
Something like:

```
git remote add TeamNewPipe https://github.com/TeamNewPipe/NewPipe.git
git fetch TeamNewPipe
git rebase --interactive TeamNewPipe/dev
```

Then close the editor it opens. After that try building it again in Android Studio and you might get some (Checkstyle) errors that you need to fix.

**Comment by werererer** [comment]
fixed it

**Comment by werererer** [comment]
is it correct that almost all variables have to be declared final now?

**Comment by wb9688** [comment]
@werererer: Yes, that's true and you could see on CI that you didn't do that. Also for the future: we prefer rebases over merges.

**Comment by werererer** [comment]
Ah sorry I messed up than :/. Fixed the checkstyle now

**Comment by wb9688** [comment]
Oh, no problem. At least CI is green now. I hope I'll be able to look at this PR soon.

**Comment by werererer** [comment]
My PR doesn't seem to be reviewed anymore :/. Should I split this PR up into multiple smaller ones so that it is easier to review? I am willing to do it if it raises the chance of getting this feature.

**Comment by wb9688** [comment]
Currently other stuff is more important, as we have a new unified player for which we have to fix the last bugs and we don't support the new Android 11 get, so those 2 have a priority to be able to get v0.20.0 out as soon as possible.

That said, could you fix the merge conflicts?

**Comment by werererer** [comment]
ah ok. Thx for clearing it up. I will fix those issues in the following days.

**Comment by Stypox** [comment]
The "seperate Drawer stuff from MainActivityClass" part is not needed anymore, as it was implemented by unified ui (now the drawer is accessible from the video detail fragment)

**Comment by werererer** [comment]
Oh I see. I have against my expectation whay less time than i expected but i will try to update this pr this or next weekend.

**Comment by opusforlife2** [comment]
> (now the drawer is accessible from the video detail fragment)

@Stypox Not if the underlying screen is a search fragment. Possibly some others too. For that we need Hashik's PR.

**Comment by werererer** [comment]
@Stypox I am not sure how the videodetailfragment relaits to the drawer. If I understand it correctly the VideodetailFragment has now the drawer implemented in it? If that is true i should probably port the important parts of my own drawerfragment class into the VideoDetailFrament class right?

**Comment by opusforlife2** [comment]
> If I understand it correctly the VideodetailFragment has now the drawer implemented in it?

No it doesn't.

**Comment by Stypox** [comment]
> > (now the drawer is accessible from the video detail fragment)
> 
> @Stypox Not if the underlying screen is a search fragment. Possibly some others too. For that we need Hashik's PR.

Which PR? Anyway yes, we need to make it so that everything is a fragment

**Comment by opusforlife2** [comment]
@Stypox #3156.

**Comment by sanchitsharma** [comment]
Is there any update on if/when this will be merged. Or is this parked due to other more prioritized work?

**Comment by litetex** [comment]
@werererer 
Thank you for the PR!
Sorry for the long review time, we are currently lacking people that review PRs.

Could you rebase and update (make sure that everything works and it fits to the app) the PR or create a new branch/PR with the changes?

**Comment by werererer** [comment]
@litetex ok i will start working on it tomorrow

**Comment by litetex** [comment]
@werererer 
Sounds great!
No stress, this PR was in such a state for a long time, so also take your time. 
It won't get deleted if it still has conflicts tomorrow 😆

Btw: I'm not sure how to handle this "bounty" label... Might have to be discussed with a more experienced dev when we merge it.

**Comment by werererer** [comment]
@litetex Ok just as an info I am currently working on it and I realized that roughly 30% of the code became too outdated. I will have to rewrite them manually so it may take a few days.

**Comment by werererer** [comment]
@litetex nevermind was easier than I though lol xD. It works now in my build I will clean now things up and make it ready to pull.

**Comment by werererer** [comment]
@litetex ok it's done :)

**Comment by litetex** [comment]
I did some further discussion with our dev team (again) and we came to the conclusion, that we don't want to add this feature as it's a) only relevant to few people (marked the issue as niche) and b) that the code - in the current form - it's not maintainable by us (would probably introduce new problems, etc).

TL;DR
The cost-benefit factor is too low.

For now I will close the PR due to the above points. Anyway, thank you for the PR and everyone else for the work that went into this.

**Comment by werererer** [comment]
ok sounds reasonable

**Comment by a-kbd** [comment]
What a shame, thank you for the hard work though @werererer
